File: apk/app/src/main/java/com/example/momentra/ui/shell/policy/ShellPolicy.kt
package com.example.momentra.ui.shell.policy

import com.example.momentra.domain.AppContext
import com.example.momentra.domain.BottomDestination
import com.example.momentra.domain.CompanySummary
import com.example.momentra.domain.MomentSummary
import com.example.momentra.domain.ShellContentState
import com.example.momentra.domain.isActiveStatus

/** Central MomentSwitcher / CompanySwitcher visibility — do not scatter in screens. */
object ShellVisibilityPolicy {
    fun showMomentSwitcher(
        context: AppContext,
        content: ShellContentState,
        destination: BottomDestination,
        activeMomentCount: Int,
        authReady: Boolean,
    ): Boolean {
        if (!authReady) return false
        if (context == AppContext.CIRCLE) return false
        if (destination == BottomDestination.CREATE) return false
        if (content !is ShellContentState.Ready && content !is ShellContentState.Empty) {
            // Empty with active moments still can show between-moments chrome; Ready required for populated.
            if (content !is ShellContentState.Empty) return false
        }
        return activeMomentCount > 0 &&
            (context == AppContext.PERSONAL || context == AppContext.GROUP || context == AppContext.BUSINESS)
    }

    fun showCompanySwitcher(context: AppContext, companies: List<CompanySummary>): Boolean =
        context == AppContext.BUSINESS && companies.isNotEmpty()
}

data class ShellInvariantInput(
    val supportedContexts: List<AppContext>,
    val selectedContext: AppContext,
    val selectedCompanyId: String?,
    val companies: List<CompanySummary>,
    val moments: List<MomentSummary>,
    val selectedMomentId: String?,
    val selectedTabByContext: Map<AppContext, BottomDestination>,
    val currentlySelectedContextDefault: AppContext = AppContext.PERSONAL,
)

data class ShellInvariantResult(
    val selectedContext: AppContext,
    val selectedCompanyId: String?,
    val selectedMomentId: String?,
    val moments: List<MomentSummary>,
    val selectedTabByContext: Map<AppContext, BottomDestination>,
    val healed: Boolean,
)

/**
 * Shell-state self-heal after bootstrap merge.
 * Business order: Context → Company → Moments(company) → Moment. Never Moment→Company.
 */
object ShellStateInvariants {
    fun heal(input: ShellInvariantInput): ShellInvariantResult {
        var healed = false
        val supported = input.supportedContexts.ifEmpty { listOf(AppContext.PERSONAL) }

        var context = input.selectedContext
        if (context !in supported) {
            context = input.currentlySelectedContextDefault.takeIf { it in supported }
                ?: supported.first()
            healed = true
        }

        var companyId = input.selectedCompanyId
        if (context != AppContext.BUSINESS) {
            if (companyId != null) {
                companyId = null
                healed = true
            }
        } else {
            val validIds = input.companies.map { it.companyId }.toSet()
            if (companyId == null || companyId !in validIds) {
                companyId = input.companies.firstOrNull()?.companyId
                healed = true
            }
        }

        val scopedMoments = when (context) {
            AppContext.BUSINESS -> input.moments.filter { m ->
                m.companyId == null || m.companyId == companyId
            }
            AppContext.CIRCLE -> emptyList()
            else -> input.moments
        }

        var momentId = input.selectedMomentId
        val momentOk = scopedMoments.any { it.momentId == momentId }
        if (momentId != null && !momentOk) {
            momentId = null
            healed = true
        }
        if (momentId == null) {
            momentId = scopedMoments.firstOrNull { it.isActiveStatus() }?.momentId
                ?: scopedMoments.firstOrNull()?.momentId
        }
        if (context == AppContext.BUSINESS && momentId != null) {
            val m = scopedMoments.firstOrNull { it.momentId == momentId }
            if (m?.companyId != null && m.companyId != companyId) {
                momentId = scopedMoments.firstOrNull { it.isActiveStatus() }?.momentId
                healed = true
            }
        }

        val tabs = input.selectedTabByContext.filterKeys { it in supported }.toMutableMap()
        for (c in supported) {
            if (c !in tabs) tabs[c] = BottomDestination.PULSE
        }

        return ShellInvariantResult(
            selectedContext = context,
            selectedCompanyId = companyId,
            selectedMomentId = momentId,
            moments = scopedMoments,
            selectedTabByContext = tabs,
            healed = healed,
        )
    }
}

enum class ShellScreenSlot {
    LOADING,
    EMPTY,
    ERROR,
    OFFLINE,
    UNAUTHORIZED,
    DEFERRED,
    PRODUCT,
    LIFE360_GLOBAL,
    PROFILE,
}

object ShellScreenResolver {
    fun resolve(
        content: ShellContentState,
        life360Open: Boolean,
        profileOpen: Boolean,
    ): ShellScreenSlot = when {
        profileOpen -> ShellScreenSlot.PROFILE
        life360Open -> ShellScreenSlot.LIFE360_GLOBAL
        content is ShellContentState.Loading || content is ShellContentState.Idle -> ShellScreenSlot.LOADING
        content is ShellContentState.Offline -> ShellScreenSlot.OFFLINE
        content is ShellContentState.Forbidden -> ShellScreenSlot.UNAUTHORIZED
        content is ShellContentState.Error -> {
            if (content.code == "UNAUTHORIZED") ShellScreenSlot.UNAUTHORIZED else ShellScreenSlot.ERROR
        }
        content is ShellContentState.Deferred -> ShellScreenSlot.DEFERRED
        content is ShellContentState.Empty -> ShellScreenSlot.EMPTY
        content is ShellContentState.Ready -> ShellScreenSlot.PRODUCT
        else -> ShellScreenSlot.LOADING
    }
}

File: momentra/momentra/Shell/Policy/ShellPolicy.swift
import Foundation

enum ShellVisibilityPolicy {
    static func showMomentSwitcher(
        context: AppContextKind,
        content: ShellContentState,
        destination: BottomDestination,
        activeMomentCount: Int,
        authReady: Bool
    ) -> Bool {
        guard authReady else { return false }
        guard context != .circle else { return false }
        guard destination != .create else { return false }
        switch content {
        case .ready, .empty: break
        default: return false
        }
        return activeMomentCount > 0 && (context == .personal || context == .group || context == .business)
    }

    static func showCompanySwitcher(context: AppContextKind, companies: [CompanySummary]) -> Bool {
        context == .business && !companies.isEmpty
    }
}

struct ShellInvariantInput {
    var supportedContexts: [AppContextKind]
    var selectedContext: AppContextKind
    var selectedCompanyId: String?
    var companies: [CompanySummary]
    var moments: [MomentSummary]
    var selectedMomentId: String?
    var selectedTabByContext: [AppContextKind: BottomDestination]
    var currentlySelectedContextDefault: AppContextKind = .personal
}

struct ShellInvariantResult {
    var selectedContext: AppContextKind
    var selectedCompanyId: String?
    var selectedMomentId: String?
    var moments: [MomentSummary]
    var selectedTabByContext: [AppContextKind: BottomDestination]
    var healed: Bool
}

enum ShellStateInvariants {
    static func heal(_ input: ShellInvariantInput) -> ShellInvariantResult {
        var healed = false
        let supported = input.supportedContexts.isEmpty ? [.personal] : input.supportedContexts

        var context = input.selectedContext
        if !supported.contains(context) {
            context = supported.contains(input.currentlySelectedContextDefault)
                ? input.currentlySelectedContextDefault
                : supported[0]
            healed = true
        }

        var companyId = input.selectedCompanyId
        if context != .business {
            if companyId != nil {
                companyId = nil
                healed = true
            }
        } else {
            let valid = Set(input.companies.map(\.companyId))
            if companyId == nil || !(valid.contains(companyId!)) {
                companyId = input.companies.first?.companyId
                healed = true
            }
        }

        let scoped: [MomentSummary]
        switch context {
        case .business:
            scoped = input.moments.filter { $0.companyId == nil || $0.companyId == companyId }
        case .circle:
            scoped = []
        default:
            scoped = input.moments
        }

        var momentId = input.selectedMomentId
        let momentOk = scoped.contains { $0.momentId == momentId }
        if momentId != nil && !momentOk {
            momentId = nil
            healed = true
        }
        if momentId == nil {
            momentId = scoped.first(where: \.isActiveStatus)?.momentId ?? scoped.first?.momentId
        }
        if context == .business, let mid = momentId,
           let m = scoped.first(where: { $0.momentId == mid }),
           let mc = m.companyId, mc != companyId {
            momentId = scoped.first(where: \.isActiveStatus)?.momentId
            healed = true
        }

        var tabs = input.selectedTabByContext.filter { supported.contains($0.key) }
        for c in supported where tabs[c] == nil {
            tabs[c] = .pulse
        }

        return ShellInvariantResult(
            selectedContext: context,
            selectedCompanyId: companyId,
            selectedMomentId: momentId,
            moments: scoped,
            selectedTabByContext: tabs,
            healed: healed
        )
    }
}

enum ShellScreenSlot {
    case loading, empty, error, offline, unauthorized, deferred, product, life360Global, profile
}

enum ShellScreenResolver {
    static func resolve(content: ShellContentState, life360Open: Bool, profileOpen: Bool) -> ShellScreenSlot {
        if profileOpen { return .profile }
        if life360Open { return .life360Global }
        switch content {
        case .loading, .idle: return .loading
        case .offline: return .offline
        case .forbidden: return .unauthorized
        case .error(let code, _):
            return code == "UNAUTHORIZED" ? .unauthorized : .error
        case .deferred: return .deferred
        case .empty: return .empty
        case .ready: return .product
        }
    }
}
